lib/Driver/ToolChains.cpp

```diff
@@ -37,8 +38,52 @@ using namespace swift;
 using namespace swift::driver;
 using namespace llvm::opt;
 
-/// The limit for passing a list of files on the command line.
-static const size_t TOO_MANY_FILES = 128;
+bool ToolChain::JobContext::shouldUseInputFileList() const {
+  if (Args.hasArg(options::OPT_driver_use_filelists))
+    return true;
+  return getTopLevelInputFiles().size() > TOO_MANY_FILES;
+}
+
+bool ToolChain::JobContext::shouldUsePrimaryInputFileList() const {
+  // SingleCompile's must not return true because then all inputs erroneously
+  // end up in primary file list.
+  if (Args.hasArg(options::OPT_driver_use_filelists))
+    return true;
+  return InputActions.size() > TOO_MANY_FILES;
+}
+
+bool ToolChain::JobContext::shouldFilterInputsByType() const {
+  // SingleCompile's must not filter inputs by type in order for the
+  // playground (test) to work.
+  return OI.CompilerMode != OutputInfo::Mode::SingleCompile;
+}
+
+bool ToolChain::JobContext::shouldUseMergeModuleInputFileList() const {
+  if (Args.hasArg(options::OPT_driver_use_filelists))
+    return true;
+  return Inputs.size() > TOO_MANY_FILES;
+}
+
+bool ToolChain::JobContext::shouldUseLinkInputFileList() const {
+  if (Args.hasArg(options::OPT_driver_use_filelists))
+    return true;
+  return Inputs.size() > TOO_MANY_FILES;
+}
+
+bool ToolChain::JobContext::shouldUseMainOutputFileList() const {
+  if (Args.hasArg(options::OPT_driver_use_filelists))
+    return true;
+  return Output.getPrimaryOutputFilenames().size() > TOO_MANY_FILES;
+}
+
+bool ToolChain::JobContext::shouldUseSupplementaryOutputFileList() const {
```

Comment:
Should this say "map"? I know the enum is still called FilelistKind or whatever, but in this case we know what we're talking about.

Reply:
This one should probably also have "frontend" in it for consistency.

Reply:
Yes, indeed! Fixed: `shouldUseSupplementaryOutputFileMapInCompilerInvocation`. Tx.